### PR TITLE
[SPARK-30754][SQL] Reuse results of floorDiv in calculations of floorMod in DateTimeUtils

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the case of back-to-back calculation of `floorDiv` and `floorMod` with the same arguments, the result of `foorDiv` can be reused in calculation of `floorMod`. The `floorMod` method is defined as the following in Java standard library:
```java
    public static int floorMod(int x, int y) {
        int r = x - floorDiv(x, y) * y;
        return r;
    }
```
If `floorDiv(x, y)` has been already calculated, it can be reused in `x - floorDiv(x, y) * y`.

I propose to modify 2 places in `DateTimeUtils`:
1. `microsToInstant` which is widely used in many date-time functions. `Math.floorMod(us, MICROS_PER_SECOND)` is just replaced by its definition from Java Math library.
2. `truncDate`: `Math.floorMod(oldYear, divider) == 0` is replaced by `Math.floorDiv(oldYear, divider) * divider == oldYear` where `floorDiv(...) * divider` is pre-calculated.

### Why are the changes needed?
This reduces the number of arithmetic operations, and can slightly improve performance of date-time functions.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
By existing test suites `DateTimeUtilsSuite`, `DateFunctionsSuite` and `DateExpressionsSuite`.

### CODE REVIEW

**Comment by srowen** [comment]
It's pretty trivially faster, I'd imagine, but the logic looks sound and it's not more complex

**Comment by kiszk** [comment]
Can we use a kind of [this optimization](https://stackoverflow.com/questions/46905740/does-java-have-a-divmod-instruction) in HotSpot?

**Comment by MaxGekk** [comment]
@kiszk The ops are not regular arithmetic `/` and `%`, they are lib calls. So, they present in byte code:
```
public microsToInstant(J)Ljava/time/Instant;
    // parameter final  us
   L0
    LINENUMBER 339 L0
    LLOAD 1
    LDC 1000000
    INVOKESTATIC java/lang/Math.floorDiv (JJ)J
   L1
    LSTORE 3
   L2
    LINENUMBER 340 L2
    LLOAD 1
    LDC 1000000
    INVOKESTATIC java/lang/Math.floorMod (JJ)J
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: For better readability, could you leave some comments about the calculation reuse here?

**Comment by maropu** [comment]
The fix looks fine to me.

**Comment by kiszk** [comment]
@MaxGekk Got it. This is not a pair of `div` and `mod`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Sorry for my naive question. Do we need `floorDiv` and `floorMod` here? In my understanding, `oldYear` and `divider` look positive.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: `oldYear` can be negative, have a look at tests in https://github.com/apache/spark/pull/25230, for instance.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I added a comment

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I see. The comment `Returns the year value for the given date. The date is expressed in days since 1.1.1970` in `getYear()` means that the value is relative based on 1.1.1970. If the year is older than 1.1.1970, the value is negative.

**Review by maropu — APPROVED** [review]
Looks fine to me. I'll leave this to @srowen and the other committers. cc: @dongjoon-hyun @HyukjinKwon

**Review by kiszk — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master (leaving out of 3.0)

**Pull request closed without merging**
